### DISCUSSION TITLE
Importing into Blender

### DISCUSSION THREAD

**Discussion by OfficialCRUGG** [BramStoutProductions/MiEx#6] [Q&A]
How would I import an export produced by MiEx into Blender? I have tried both just importing the "main" file, or all the files one-by-one (since Blender doesn't seem to allow importing multiple files at once)
I get the geometry, but "no materials" (there are materials, but they're all just solid gray; see images below)
![image](https://github.com/BramStoutProductions/MiEx/assets/25248999/3b114098-0a9d-41da-8c2c-4a35b42a01ce)
![image](https://github.com/BramStoutProductions/MiEx/assets/25248999/3b799973-74aa-486a-b6b8-71235a32a02c)

**Comment by zNightlord** [comment]
> How would I import an export produced by MiEx into Blender? I have tried both just importing the "main" file, or all the files one-by-one (since Blender doesn't seem to allow importing multiple files at once) I get the geometry, but "no materials" (there are materials, but they're all just solid gray; see images below) ![image](https://private-user-images.githubusercontent.com/25248999/325022598-3b114098-0a9d-41da-8c2c-4a35b42a01ce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM5NTM1MTIsIm5iZiI6MTcxMzk1MzIxMiwicGF0aCI6Ii8yNTI0ODk5OS8zMjUwMjI1OTgtM2IxMTQwOTgtMGE5ZC00MWRhLThjMmMtNGEzNWI0MmEwMWNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI0VDEwMDY1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVjOTY4NGI4MTIyOTQyOTE4YWRjMjA2MjQzMTgwOWE2ZjIyNmEwYmZlYTkxODhlNzU2Y2M2ZjdhMTYwODQ1ZjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xSdYDQy48ZRj0VLw22nvyc1QE2qlaq_1GEAeAcXO6zQ) ![image](https://private-user-images.githubusercontent.com/25248999/325022617-3b799973-74aa-486a-b6b8-71235a32a02c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM5NTM1MTIsIm5iZiI6MTcxMzk1MzIxMiwicGF0aCI6Ii8yNTI0ODk5OS8zMjUwMjI2MTctM2I3OTk5NzMtNzRhYS00ODZhLWI2YjgtNzEyMzVhMzJhMDJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI0VDEwMDY1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc4M2RmYmQxYTU0NDc0NWM2NDQ3YzgyNzNjYzVjNDA0ODUyNjg5MGJjZjY5ZTkwMWNkZDM2NDIyNTJkYTA1OGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oozct91LXG2sfpPbsN_Faja7ZBV5aVYsJ-XW7fTFV54)

First of all welcome to USD.
I haven't tried the exporter personally yet. 

This should give you the rough idea of what to do. 
You see Blender hasn't fully support a system called MaterialX through the builtin importer and the exporter use MaterialX (by default?) What you need to do is use UsdPreviewSurface. See templates.

https://github.com/BramStoutProductions/MiEx/wiki/6.-USD#material-templates
https://github.com/BramStoutProductions/MiEx/blob/main/extras/example_resource_packs/UsdPreviewSurface/materials/minecraft/templates/base.json

**Reply by AstralRandomness** [reply]
How would that work? I'm also new to USD, and unsure of how exactly I can use UsdPreviewSurface even after having read the documentation.

**Reply by BramStout** [reply]
Blender does indeed not support MaterialX. MiEx by default uses MaterialX because that allows it to do biome colours and animated textures.

In the repository you do have an example resource pack called `UsdPreviewSurface`, this will cause MiEx to export the materials out as UsdPreviewSurface materials which Blender can import in. However, you then don't have biome colours or animated texture. So you'd still need to fix up the material in Blender after importing.

You can use the `UsdPreviewSurface` resource pack really easily. If you download the repo, then all you need to do is copy that the `UsdPreviewSurface` folder into the `resources` folder that's next to the jar file. Then in MiEx you can enable the resource pack on the right part of the window.

**Reply by zNightlord** [reply]
> you then don't have biome colours or animated texture. So you'd still need to fix up the material in Blender after importing.

As for this, if someone interested in solving this problem. I recommend looking at
[USDHook - Blender 4.1](https://docs.blender.org/api/4.1/bpy.types.USDHook.html)

**Comment by BlueEvilGFX** [comment]
A workaround for having correct materials when importing `MiEx` worlds into blender would be this:
- download the `miex_helper.py` add-on from the `MiEx` repository [here](https://github.com/BramStoutProductions/MiEx/blob/main/extras%2FBlender%2FBlenderAddon%2FMiEx_helper.py)
- install the add-on
- after the `.usd` import select the `world empty` and look in the search bar (in the newer blender version the default hotkey for this is `f3`) for `miex.prep`
- prep the materials with this operator

Disclaimers:
The add-on is just a quick fix and should be improved when more `MiEx` utilities are added in...

**Reply by AstralRandomness** [reply]
Quick question, I got the addon but I can't find anything relating to 'world Empty' anywhere. where is it supposed to be?

**Reply by BlueEvilGFX** [reply]
When importing the world into blender (with usd previews used for exporting in MiEx), it will create empties which control the mc chunks in blender. You select the top level empty, which is i guess the name of the usd file. An example for this can be seen in the following screenshot.
![image](https://github.com/BramStoutProductions/MiEx/assets/107540992/8e3b1182-9a13-4242-8b4e-33bd893b118e)

**Reply by AstralRandomness** [reply]
I've tried, and unfortunately I still can't figure it out. Not sure what I did wrong - I have the world exported correctly, the texturepack loaded in MiEx, and the plugin on - but it still doesn't show up when I select and search for the command. If it matters, I'm using Blender 4.1.1 instead of 4.1.0. You can see the issue in the attached screenshot.
![image](https://github.com/BramStoutProductions/MiEx/assets/78801245/60200760-1a0c-43b9-8364-ac8bc96604a5)

**Reply by BlueEvilGFX** [reply]
I can sadly not reproduce the error on blender version `4.1.1`. So please send me the error message of the blender system console if available. Thank you very much!

**Reply by AstralRandomness** [reply]
It is indeed available. It's attached here - the whole thing. My username is censored fyi

```
Read prefs: "C:\Users\USERNAME\AppData\Roaming\Blender Foundation\Blender\4.1\config\userpref.blend"
Failed to register scene update handler
        addon_utils.reset_all unloading node_arrange
Info: Deleted 1 object(s)

USD import of 'C:\Users\USERNAME\Downloads\usd export from miex\rpggame-testexport.usd' took 1001.30 ms
addon_utils.disable: MiEx_helper not enabled
addon_utils.disable: MiEx_helper not enabled
Modules Installed (MiEx_helper) from 'C:\\Users\\USERNAME\\Downloads\\MiEx_helper.py' into 'C:\\Users\\USERNAME\\AppData\\Roaming\\Blender Foundation\\Blender\\4.1\\scripts\\addons'
module changed on disk: 'C:\\Users\\USERNAME\\AppData\\Roaming\\Blender Foundation\\Blender\\4.1\\scripts\\addons\\MiEx_helper.py' reloading...
Error: Failed to find 'prepmiex'
USD import of 'C:\Users\USERNAME\Downloads\usd export from miex\rpggame-testexport_materials.usd' took 205.65 ms
```

**Reply by BlueEvilGFX** [reply]
Just for clarification: the Error appears not in the log when the addon is enabled, but when you try to prep the world?

I don't think this will solve the issue, but for importing and using the world in blender, you need to click on the `+` in Miex to use `UsdPreviewSurface`, as you can see in the screenshot below.
If you look for only `miex` in the search bar, it does't show up either, does it?

![image](https://github.com/BramStoutProductions/MiEx/assets/107540992/baad25f1-95cc-40ef-aadf-7290dbfd486f)

**Reply by AstralRandomness** [reply]
The error appears when I try and prep the world. The addon enabling (or disabling) works fine. If I search for MiEx, it does not show up either. Also, I'll try the USDPreviewSurface again to see if it solves the issue or makes it more apparent.

**Reply by BlueEvilGFX** [reply]
This confuses me. You said that you cannot prep the world since the operator does not show up. So the error message shows up when you open the search bar then? But not otherwise?

**Reply by AstralRandomness** [reply]
I was confused myself when writing the reply above yours. Let me correct myself. The error message appears only after I try and search for the operator.

**Reply by BlueEvilGFX** [reply]
I might have found the culprit for this. Please activate the `python tooltips` in the blender preferences. Then try again prepping the world. (In `MiEx` you need to use `UsdPreviewSurface`)
Please then reply if this has solved the issue or not. If so, I can fix this with the next pull request for the addon.

![image](https://github.com/BramStoutProductions/MiEx/assets/107540992/2e53223d-9ab7-409e-ae2f-608986c6eaa1)

**Reply by AstralRandomness** [reply]
It worked! Thank you so much and I am so so sorry for the repeated issues and confusion.

**Reply by BlueEvilGFX** [reply]
I am sorry myself that a tiny amount of lines missing in the code created this issue. So I am glad that you reported it, no matter how confusing it was in the end. Thank you so much! :)
At the moment, I prepare an update to the addon which should be in a pull request in the next few 10min, until BramStout merges the pull request. Thank you very much and I am happy that it works for you now! 😊

**Reply by AstralRandomness** [reply]
Just a small question, is it supposed to automatically set up the materials after it sets up the world, or do I need to import the material USD and manually assign the materials

**Reply by BlueEvilGFX** [reply]
You do not need to import the `materials.usd`. You select the `world` empty, then prep the world. It will automatically set up the materials for you.

**Reply by AstralRandomness** [reply]
Ah, I see. In that case, something is still wrong. If I prep the world, it works just fine - but stays the default material.

**Reply by BlueEvilGFX** [reply]
Please provide a screenshot of your descirbed issue, thanks! If the world is fully white, then you have forgot to choose export with `UsdMaterialPreview` in the exporter application MiEx

**Reply by AstralRandomness** [reply]
I attached a screenshot below. The world got prepped and made smaller, but the materials are white. This might be because I didn't download USDPreviewSurface correctly. I will try fixing the issue and will update with results (or lack thereof)
![image](https://github.com/BramStoutProductions/MiEx/assets/78801245/51bf9701-5c8c-4edb-afd9-210a9abbe710)
